examples/fluids/qfunctions/advection2d.h

```diff
@@ -448,6 +447,57 @@ CEED_QFUNCTION(Advection2d_Sur)(void *ctx, CeedInt Q,
   const CeedScalar strong_form = context->strong_form;
   const bool implicit          = context->implicit;
 
+  CeedPragmaSIMD
+  // Quadrature Point Loop
+  for (CeedInt i=0; i<Q; i++) {
+    // Setup
+    // -- Interp in
+    const CeedScalar rho        =    q[0][i];
+    const CeedScalar u[3]       =   {q[1][i] / rho,
+                                     q[2][i] / rho,
+                                     q[3][i] / rho
+                                    };
+
+    // -- Interp-to-Interp q_data
+    // For explicit mode, the surface integral is on the RHS of ODE q_dot = f(q).
+    // For implicit mode, it gets pulled to the LHS of implicit ODE/DAE g(q_dot, q).
+    // We can effect this by swapping the sign on this weight
+    const CeedScalar wdetJb     =   (implicit ? -1. : 1.) * q_data_sur[0][i];
+
+    // ---- Normal vectors
+    const CeedScalar norm[2]    =   {q_data_sur[1][i],
+                                     q_data_sur[2][i]
+                                    };
+    // Normal velocity
+    const CeedScalar u_normal = norm[0]*u[0] + norm[1]*u[1];
+
+    // No Change in density or momentum
+    for (CeedInt j=0; j<4; j++)
+      v[j][i] = 0;
+
+    // Implementing inflow BC
+    if (u_normal < 0)
```

Comment:
is this `if` needed here?

Reply:
Good point. For advection and advection2d, I prefer to leave the if-statement since it's safer to define inflow and outflow by the u_normal. In advection2d, it is easy to figure out the inflow and outflow face IDs by looking at the velocity components but it becomes tedious/tricky in advection.

Reply:
Is it ok to possibly end up in a state where v[0..3] are changed but not v[4]?

Reply:
I had the same thought and tried to zero v[4] before the if-statement but it didn't affect the results. I can zero it again to be safe.

Reply:
I don't understand what splitting this QFunction and repeating most of it gives us?

Reply:
This setup works for advection and euler but when we have viscosity (Navier-Stokes) `u_normal` is not a proper criterion in defining the in/outflow faces anymore. In the presence of viscosity, the user should give us the inflow and outflow face IDs. By splitting this QFunction, I was mostly envisioning a generic design for all our fluid problems, especially the one that @jrwrigh is working on. For turbulence simulation, we would need to treat the inflow and outflow boundaries differently.

Reply:
That makes sense. My concern here though is we have taken 1 QFunction with 2 possible outcomes and converted it into 2 QFunctions with 3 (or 4?) possible outcomes. Was this the intention?

Reply:
I think it'd be best to either leave it in one function or to error in case an inflow isn't an inflow. It's likely that a user will want to make an inflow condition with different (perhaps time-dependent) profile, but probably always the same outflow condition.

Reply:
@LeilaGhaffari, you can return a non-zero value from a QFunction and on the CPU side this will get `CeedChk`ed and stop the code.

Reply:
Ok, I just rebased this branch afresh and combined the inflow and outflow BCs for Advection(2d). We can change this later. I am seeing the original error that I reported on Zulip. Advection and Advection2d are failing for similar errors. Other tests are passing.
![Screenshot from 2022-01-29 16-57-09](https://user-images.githubusercontent.com/49916147/151681584-17319b45-adb2-4332-a838-e7dcf5f32fd5.png)

Reply:
> @LeilaGhaffari, you can return a non-zero value from a QFunction and on the CPU side this will get `CeedChk`ed and stop the code.

@jeremylt , sorry, I am not following. What is this for?

Reply:
Does the above run correctly locally?

Reply:
No, same issue on my machine.

Reply:
Returning nonzero is the way to signal an error, as when a boundary designated as inflow receives an outflow velocity.

Reply:
I pushed a fix for the missing `DMPlexLabelComplete`, which happens automatically when there is any essential boundary condition defined over `"Face Sets"`. But I don't think it should be required here. Specifically, `DMPlexGetLocalOffsets` calls
```c
  if (domain_label) {
    ierr = DMGetFirstLabelEntry_Internal(dm, dm, domain_label, 1, &label_value, dim, NULL, &ds);CHKERRQ(ierr);
  }
```
which effectively requires that a vertex is returned by `DMGetEnclosurePoint`:
```c
    ierr = DMLabelGetStratumIS(label, ids[i], &pointIS);CHKERRQ(ierr);
    if (!pointIS) continue; /* No points with that id on this process */
    ierr = ISGetIndices(pointIS, &points);CHKERRQ(ierr);
    ierr = DMGetEnclosurePoint(dm, odm, enc, points[0], &point);CHKERRQ(ierr);
    ierr = DMLabelGetValue(depthLabel, point, &pdepth);CHKERRQ(ierr);
    if (pdepth == cdepth) {
      ls = point;
      if (ds) {ierr = DMGetCellDS(dm, ls, ds);CHKERRQ(ierr);}
    }
```

@knepley What would be the correct logic here so that a height 1 label does not need to be propagated to vertices?

Reply:
I am not understanding the question. `DMGetFirstLabelEntry_Internal` tries to return a point to you of the height that you ask for in the call. `DMGetEnclosurePoint` is just trying to handle asking for points in related submeshes.

Reply:
I need this to handle both cases, where the label has been completed and when it hasn't. Is there a better way? It's kind of wasteful to do these hefty intersections for every boundary.
https://gitlab.com/petsc/petsc/-/merge_requests/4786

Reply:
As long as you are always asking for a face, it should not matter whether the label has been completed.

Reply:
Thanks for the fix @jedbrown . You have probably noticed that `euler_vortex` is now failing.

Reply:
I think it's clean with my PETSc branch above. If @knepley approves, I'll document and get that branch merged quickly. We don't have to wait for it to merge this branch, but I'd like to know it's okay in principle before proceeding here.